src/main/scala/reactor/Dispatcher.scala

```diff
@@ -1,30 +1,83 @@
-// group 1
-// 123456 Firstname Lastname
-// 654321 Firstname Lastname
+// group 17
+// 906984 Yerzhan Zhamashev
+// 936323 Nikolai Semin
 
 package reactor
 
-import reactor.api.{Event, EventHandler}
+import reactor.api.{Event, EventHandler, Handle}
+import scala.collection.mutable.Set
+import scala.collection.mutable.ListBuffer
 
 final class Dispatcher(private val queueLength: Int = 10) {
   require(queueLength > 0)
 
+  private val handlers: Set[EventHandler[Any]] = Set()
+  private val eventsQueue = new BlockingEventQueue[Any](queueLength)
+  private val threadList: ListBuffer[WorkerThread] = ListBuffer.empty
+
+  /* The idea would be to start a new thread for each added handler monitoring events in there.
+  Once a handle is removed the corresponding thread should be interrupted.
+   */
   @throws[InterruptedException]
-  def handleEvents(): Unit = ???
+  def handleEvents(): Unit = {
+    while (handlers.nonEmpty) {
+      val event = eventsQueue.dequeue
+      val handler = event.getHandler
+      if (handlers.contains(handler)){
+        event.dispatch()
+      }
+    }
+  }
 
-  def addHandler[T](h: EventHandler[T]): Unit = ???
+  def addHandler[T](h: EventHandler[T]): Unit = {
+    val thread = new WorkerThread(h.asInstanceOf[EventHandler[Any]])

```

Comment:
Would be good to check if the same hanlder exists already before starting a new thread